feat(ui): show timestamp in flow revision dropdown

Suraj-kumar00 opened the pull request:
### ✨ Description

This PR adds timestamps to the flow revision dropdown, making it easier to identify and distinguish between past revisions.

**Changes:**
- **Backend:** Added database migration [`V1_51__flows_updated_date.sql`](kestra/jdbc-h2/src/main/resources/migrations/h2/V1_51__flows_updated_date.sql:0:0-0:0) to add `updated` timestamp column to flows table, and updated `AbstractJdbcFlowRepository.findRevisions()` to return the timestamp with each revision
- **Frontend:** Updated [`FlowRevisions.vue`](kestra/ui/src/components/flows/FlowRevisions.vue:0:0-0:0) to load actual revision data from API (with timestamps), and updated [`Revisions.vue`](kestra/ui/src/components/layout/Revisions.vue:0:0-0:0)](kestra/ui/src/components/layout/Revisions.vue:0:0-0:0) with `YYYY-MM-DD HH:mm` format, custom el-option slots, and styled display (grey color, right-aligned)

**Result:** Revision dropdown now shows `6 (current)` with `2025-12-21 00:45` right-aligned in grey.

### 🔗 Related Issue

Closes https://github.com/kestra-io/kestra/issues/13669

### 🎨 Frontend Checklist

- [x] Code builds without errors (`npm run build`)
- [x] All existing E2E tests pass (`npm run test:e2e`)
- [x] Screenshots or video recordings attached showing the `UI` changes

**Screenshot:**
<img width="1512" height="911" alt="Screenshot 2026-01-05 at 11 09 19 PM" src="https://github.com/user-attachments/assets/bedbeb75-8066-450e-a42e-c05573eac769" />


### 🛠️ Backend Checklist

- [x] Code compiles successfully and passes all checks
- [x] Unit test shouldReturnUpdatedDateInFindRevisions() passes
- [ ] All unit and integration tests pass

### 📝 Additional Notes

- Existing revisions created before this migration will show the migration timestamp (when `ALTER TABLE` ran)
- Only new revisions will have accurate creation timestamps
- The timestamp format `YYYY-MM-DD HH:mm` matches the format requested in the issue

- ✅ Use moment library for date formatting 
- ✅ Don't load revisions when already in store
- ✅ Move updatedDate to AbstractFlow.java 
- ✅ Add H2 and MySQL migrations
- ✅ Add Java unit tests

**Latest Updates:**
- Fixed jOOQ `Record4` → `Record5` type to match selected columns including `updated`
- Updated Storybook test expectations for new dropdown format
- Fixed backend test to use updated findRevisions() method signature

Suraj-kumar00 committed:
feat(ui): show timestamp in flow revision dropdown (#13669)

- Add database migration to track flow revision timestamps
- Update backend to return updatedDate in revision API response
- Display timestamps in YYYY-MM-DD HH:mm format in dropdown
- Style timestamps with grey color and right alignment

Closes #13669

Suraj-kumar00 commented:
Hi @Skraye,

Please review!

Skraye commented on ui/src/components/layout/Revisions.vue:
```diff
@@ -216,10 +227,47 @@
+        const year = date.getFullYear();
+        const month = String(date.getMonth() + 1).padStart(2, "0");
+        const day = String(date.getDate()).padStart(2, "0");
+        const hours = String(date.getHours()).padStart(2, "0");
+        const minutes = String(date.getMinutes()).padStart(2, "0");
+        
+        return `${year}-${month}-${day} ${hours}:${minutes}`;
+    }
```
Please use the moment library

Skraye commented on ui/src/components/flows/FlowRevisions.vue:
```diff
@@ -26,12 +26,29 @@
+            });
+            revisions.value = data;
+        }
+    });
```
Dont loadRevisions when we already have them in the flow object

Skraye commented on core/src/main/java/io/kestra/core/models/flows/FlowWithSource.java:
```diff
@@ -16,6 +17,8 @@
 
     String source;
+    
+    Instant updatedDate;
```
Put the updatedDate inside the `AbstractFlow.java` please, as its there that we have the revision prop at first

Skraye commented on jdbc-postgres/src/main/resources/migrations/postgres/V1_53__flows_updated_date.sql:
```diff
@@ -0,0 +1,10 @@
+-- Create trigger to auto-update timestamp on each insert/update
+CREATE OR REPLACE TRIGGER flows_updated BEFORE INSERT OR UPDATE
+    ON flows FOR EACH ROW EXECUTE PROCEDURE
+    UPDATE_UPDATED_DATETIME();
```
Add the same for h2 and mysql please

Skraye replied:
```suggestion
alter table flows add "updated" VARCHAR(250) GENERATED ALWAYS AS (value ->> 'updated') STORED;
```

Skraye reviewed (changes requested):
Solid base but need some changes.
Also please provide some Java tests to validate your changes.

Ping me if you need help for something!

Suraj-kumar00 committed:
fix: address PR review feedback for flow revision timestamps

- Use moment library instead of manual date formatting (Revisions.vue)
- Avoid duplicate API calls by using existing store data (FlowRevisions.vue)
- Move updatedDate field from FlowWithSource to AbstractFlow
- Add database migrations for H2 (V1_53) and MySQL (V1_54)
- Fix PostgreSQL migration version (V1_51 → V1_53)
- Add unit test shouldReturnUpdatedDateInFindRevisions()

Suraj-kumar00 committed:
feat: restore timestamp display alongside delete button in revision dropdown

Suraj-kumar00 committed:
fix: update test expectations for new revision dropdown format                                 ─╯

- Update Storybook test to expect 'Revision N' instead of 'N'
- Fix findRevisions call to use 4 parameters (added allowDeleted)

Suraj-kumar00 commented:
Hi @Skraye, how have you been?

Please review and let me know if anythings required from my end.

Also tried fixing the CI failures but the backend CI is failing due to this:

>  17:23:06.684 ERROR Test worker  io.kestra.core.secret.SecretService Could not decode secret 'SECRET_NON_B64_SECRET', make sure it is Base64-encoded: Illegal base64 character 20

Suraj-kumar00 commented:
Hi @Skraye, please review!

Skraye commented on jdbc-h2/src/main/resources/migrations/h2/V1_53__flows_updated_date.sql:
```diff
@@ -0,0 +1,5 @@
+ALTER TABLE flows ADD COLUMN IF NOT EXISTS "updated" TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP;
+
+-- Create index for efficient ordering by timestamp
+CREATE INDEX IF NOT EXISTS flows_updated ON flows ("updated");
```
```suggestion
alter table flows add "updated" VARCHAR(250) GENERATED ALWAYS AS (JQ_STRING("value", '.updated'));
```

Skraye commented on jdbc/src/main/java/io/kestra/jdbc/repository/AbstractJdbcFlowRepository.java:
```diff
@@ -309,17 +309,23 @@ public List<FlowWithSource> findRevisions(String tenantId, String namespace, Str
                     .and(field("id", String.class).eq(id))
                     .orderBy(REVISION_FIELD.asc());
 
                 return select.fetch()
```
This isnt necessary, the value field will return the whole object containing the updated property, please remove it

Skraye commented on core/src/main/java/io/kestra/core/models/flows/AbstractFlow.java:
```diff
@@ -37,6 +37,9 @@ public abstract class AbstractFlow implements FlowInterface {
     Integer revision;
 
+    @Schema(description = "The timestamp when this revision was created or last updated.")
+    java.time.Instant updatedDate;
```
```suggestion
    java.time.Instant updated;
```

Skraye commented on core/src/main/java/io/kestra/core/models/flows/FlowWithSource.java:
```diff
@@ -87,4 +88,33 @@ public static FlowWithSource of(Flow flow, String source) {
+            .updatedDate(updatedDate)
+            .concurrency(flow.concurrency)
+            .retry(flow.retry)
+            .sla(flow.sla)
+            .checks(flow.checks)
+            .build();
+    }
 }
```
Property updated is in flow as it comes from `AbstractFlow` so no need for a new method
```suggestion
            .checks(flow.checks)
            .updated(flow.updated)
            .build();
    }
    
}
```

Skraye reviewed (changes requested):
Hello

- For the SQL in fact we never store the property directly, instead we store the object as JSON, and we create column based on this json (see suggested changes)

- Java Test is not passing, I also notice you have no places where you set the updatedDate

Suraj-kumar00 committed:
fix(ui): add timestamps to flow revision dropdown

- Add  property to AbstractFlow to track revision creation time
- Store updated timestamp when saving flows in AbstractJdbcFlowRepository
- Add SQL migrations using GENERATED ALWAYS AS pattern (H2, PostgreSQL, MySQL)
- Display timestamps in revision dropdown in Revisions.vue
- Remove redundant FlowWithSource.of() overload per maintainer feedback
- Simplify findRevisions() - value JSON already contains updated property

Fixes #13782

Suraj-kumar00 commented:
Hi @Skraye, 

Please review.

MilosPaunovic commented:
Hi @Suraj-kumar00,

This pull request will be reviewed at our earliest convenience.

Thanks for all the efforts so far! 🍀

Skraye committed:
fix: PR polishing

Skraye reviewed (approved):
LGTM

Skraye committed:
Merge branch 'develop' into feature/ui-revision-timestamps-13669

Skraye committed:
fix: tests

Suraj-kumar00 committed:
Merge branch 'develop' into feature/ui-revision-timestamps-13669
